File: scripts/fetch_avatar.py
#!/usr/bin/env python3
"""Retrieve source-pinned classroom portrait; verify Git blob identity and GLB header."""
import base64
import hashlib
import json
from pathlib import Path
import urllib.request
ROOT=Path(__file__).resolve().parents[1]
SPEC=json.loads((ROOT/'frontend/public/avatars/ce-brunette/SOURCE.json').read_text())
def identity(data):return hashlib.sha1(b'blob '+str(len(data)).encode()+b'\0'+data).hexdigest()
def main():
    dst=ROOT/'frontend/public/avatars/ce-brunette/avatar.glb';dst.parent.mkdir(parents=True,exist_ok=True)
    if dst.exists() and identity(dst.read_bytes())==SPEC['git_blob_sha']:return
    for url in [SPEC['url'],'https://api.github.com/repos/met4citizen/TalkingHead/git/blobs/'+SPEC['git_blob_sha']]:
        try:
            print('AVATAR_SOURCE '+url,flush=True)
            request=urllib.request.Request(url,headers={'User-Agent':'Socrates-Run2'})
            with urllib.request.urlopen(request,timeout=45) as r:data=r.read(12000000)
            if 'api.github.com' in url:data=base64.b64decode(json.loads(data)['content'])
            if identity(data)!=SPEC['git_blob_sha'] or data[:4]!=b'glTF':raise ValueError('ASSET_IDENTITY')
            dst.write_bytes(data);return
        except Exception as e:print('AVATAR_ATTEMPT '+type(e).__name__,flush=True)
    raise RuntimeError('PINNED_AVATAR_SOURCE_AWAITING')
if __name__=='__main__':main()

File: frontend/tools/fetch-avatar.mjs
import fs from 'node:fs';import crypto from 'node:crypto';
const dir='public/avatars/ce-brunette',spec=JSON.parse(fs.readFileSync(dir+'/SOURCE.json','utf8'));
const hash=b=>crypto.createHash('sha1').update(Buffer.from('blob '+b.length+'\0')).update(b).digest('hex');
const path=dir+'/avatar.glb';
if(!(fs.existsSync(path)&&hash(fs.readFileSync(path))===spec.git_blob_sha)){
  let found=false;
  for(const url of [spec.url,'https://api.github.com/repos/met4citizen/TalkingHead/git/blobs/'+spec.git_blob_sha]){
    try{const r=await fetch(url,{headers:{'User-Agent':'Socrates-Run2'},signal:AbortSignal.timeout(45000)});if(!r.ok)throw Error('HTTP '+r.status);
      const b=url.includes('api.github.com')?Buffer.from((await r.json()).content,'base64'):Buffer.from(await r.arrayBuffer());
      if(hash(b)!==spec.git_blob_sha||b.subarray(0,4).toString()!=='glTF')throw Error('ASSET_IDENTITY');fs.writeFileSync(path,b);found=true;break;
    }catch(e){console.error('AVATAR_ATTEMPT',e.name)}
  }
  if(!found)throw Error('PINNED_AVATAR_SOURCE_AWAITING');
}
